src/daft-io/src/object_io.rs

```diff
@@ -112,10 +117,203 @@ pub(crate) trait ObjectSource: Sync + Send {
+                        ),
+                    }
+                // Last fragment does not contain wildcard: we return it if the full path exists and is a FileType::File
+                } else {
+                    let full_dir_path = path.to_string() + current_fragment.escaped_str().as_str();
+                    let single_file_ls = source.ls(full_dir_path.as_str(), Some("/"), None).await;
+                    match single_file_ls {
+                        Ok(mut single_file_ls) => {
+                            if single_file_ls.files.len() == 1
+                                && matches!(single_file_ls.files[0].filetype, FileType::File)
+                            {
+                                let fm = single_file_ls.files.drain(..).next().unwrap();
+                                result_tx.send(Ok(fm)).await.expect("Internal multithreading channel is broken: results may be incorrect");
+                            }
+                        }
+                        Err(super::Error::NotFound { .. }) => (),
+                        Err(e) => result_tx.send(Err(e)).await.expect(
+                            "Internal multithreading channel is broken: results may be incorrect",
+                        ),
+                    };
+                }
+
+            // RECURSIVE CASE: current_fragment contains a special character (e.g. *)
+            } else if current_fragment.has_special_character() {
+                let partial_glob_matcher =
+                    GlobBuilder::new(GlobFragment::join(&glob_fragments[..i + 1], "/").raw_str())
+                        .literal_separator(true)
+                        .build()
+                        .expect("Cannot parse glob")
+                        .compile_matcher();
+                let next_level_file_metadata =
+                    source.iter_dir(path.as_str(), Some("/"), None).await;
+
+                match next_level_file_metadata {
+                    Ok(mut next_level_file_metadata) => {
+                        while let Some(fm) = next_level_file_metadata.next().await {
+                            match fm {
+                                Ok(fm) => {
+                                    if matches!(fm.filetype, FileType::Directory) && partial_glob_matcher.is_match(fm.filepath.as_str().trim_end_matches('/')) {
+                                        visit(
+                                            result_tx.clone(),
+                                            source.clone(),
+                                            fm.filepath.as_str(),
+                                            (glob_fragments.clone(), i + 1),
+                                            full_glob_matcher.clone(),
+                                        );
+                                    }
+                                }
+                                Err(super::Error::NotFound { .. }) => (),
```

Comment:
Why are we surpressing the NotFound?

Reply:
This is because of the following possible scenario:

Folder structure:
```
/
  x/
    a/
      foo.txt
  y/
    b/
      foo.txt
```

When we call `io_glob("/*/a/foo.txt")`, one of the intermediate calls will be calling a list on `y/a/` as it recurses down the tree. If we don't suppress the NotFound, the overall call will throw a NotFound error.